src/System.Runtime.Extensions/src/System/Environment.CoreCLR.cs

```diff
@@ -0,0 +1,406 @@
+
+        private static string ExpandEnvironmentVariablesCore(string name)
+        {
+            int currentSize = 100;
+            StringBuilder result = StringBuilderCache.Acquire(currentSize); // A somewhat reasonable default size
+
+            result.Length = 0;
+            int size = Interop.mincore.ExpandEnvironmentStringsW(name, result, currentSize);
+            if (size == 0)
+            {
+                StringBuilderCache.Release(result);
+                Marshal.ThrowExceptionForHR(Marshal.GetHRForLastWin32Error());
+            }
+
+            while (size > currentSize)
+            {
+                currentSize = size;
+                result.Capacity = currentSize;
+                result.Length = 0;
+
+                size = Interop.mincore.ExpandEnvironmentStringsW(name, result, currentSize);
+                if (size == 0)
+                {
+                    StringBuilderCache.Release(result);
+                    Marshal.ThrowExceptionForHR(Marshal.GetHRForLastWin32Error());
+                }
+            }
+
+            return StringBuilderCache.GetStringAndRelease(result);
+        }
+
+        private static string GetEnvironmentVariableCore(string variable, EnvironmentVariableTarget target)
+        {
+            if (target == EnvironmentVariableTarget.Process)
+            {
+                StringBuilder sb = StringBuilderCache.Acquire(128); // a somewhat reasonable default size
+                int requiredSize = Interop.mincore.GetEnvironmentVariableW(variable, sb, sb.Capacity);
+                if (requiredSize == 0 && Marshal.GetLastWin32Error() == Interop.mincore.Errors.ERROR_ENVVAR_NOT_FOUND)
+                {
+                    StringBuilderCache.Release(sb);
+                    return null;
+                }
+
+                while (requiredSize > sb.Capacity)
+                {
+                    sb.Capacity = requiredSize;
+                    sb.Length = 0;
+                    requiredSize = Interop.mincore.GetEnvironmentVariableW(variable, sb, sb.Capacity);
+                }
+
+                return StringBuilderCache.GetStringAndRelease(sb);
+            }
+            else if (target == EnvironmentVariableTarget.Machine)
+            {
+                // TODO #8533: Uncomment/fix when registry APIs available
+                //using (RegistryKey environmentKey = Registry.LocalMachine.OpenSubKey(@"System\CurrentControlSet\Control\Session Manager\Environment", false))
+                //{
+                //    return environmentKey?.GetValue(variable) as string ?? null;
+                //}
+                return null;
+            }
+            else
+            {
+                Debug.Assert(target == EnvironmentVariableTarget.User);
+
+                // TODO #8533: Uncomment/fix when registry APIs available
+                //using (RegistryKey environmentKey = Registry.CurrentUser.OpenSubKey("Environment", false))
+                //{
```

Comment:
This implementation is not going to be tree-shaking friendly. If I am just using process environment (the common case), it is going to pull in the bag of rare registry stuff. When we see things like this, we may want to factor them to be tree-shaking friendly. In this case, it would mean avoiding the registry in the transitive closure of the process APIs (it would require a few more methods in the PAL).

Mono BCL tends to have tweaks like this to make their BCL tree-shake well.